### PR TITLE
Fix debug information

### PR DESCRIPTION
**Pull request by MartinNowack**
@hanouchen As disccussed, here are the patches which should give you back the power of the debug instructions. Especially, have a look at https://github.com/hanouchen/klee/commit/cb0443d80dfd2cf34bba5d9ae6248fd2b4a8a7ec which gives you a detailled idea for iterating over the debug instructions. This way you should be able to extract most of the information. In case of any questions, let me know.

### CODE REVIEW

**Comment by hanouchen** [comment]
@MartinNowack Thanks Martin!

**Comment by hanouchen** [comment]
@MartinNowack 
I got these errors when compiling the code:
screen shot 2018-05-05 at 11 23 16
I'm using clang & llvm version 3.4, following [this](http://klee.github.io/build-llvm34/) build guide, any idea what's causing the problem here? Thanks.

File: include/klee/Internal/Module/InstructionInfoTable.h [hunk 1]
```diff
@@ -14,10 +14,20 @@
 #include <string>
 #include <set>
 
+#include <ciso646>
```
Comment (MartinNowack) [inline include/klee/Internal/Module/InstructionInfoTable.h:17]: Remove the following lines until the closing '#endif'

File: include/klee/Internal/Module/InstructionInfoTable.h [hunk 2]
```diff
@@ -50,13 +56,15 @@ namespace klee {
 
     std::string dummyString;
     InstructionInfo dummyInfo;
-    std::map<const llvm::Instruction*, InstructionInfo> infos;
+    unordered_map<const llvm::Instruction *, InstructionInfo> infos;
```
Comment (MartinNowack) [inline include/klee/Internal/Module/InstructionInfoTable.h:59]: add a `std::` in front of it and the following line.
This should fix your compiler issues

**Comment by MartinNowack** [comment]
@hanouchen I added some comments in a nutshell remove the `definition` block from the beginning of the file and use the `std::unordered_map` for the rest.

**Comment by hanouchen** [comment]
@MartinNowack Working now, thanks.

**Comment by hanouchen** [comment]
@MartinNowack Hey Martin, do you still have the small program you used to demonstrate the debug intrinsics during the project meeting last week? Just want to play around and experiment with it if that's ok :)

**Comment by MartinNowack** [comment]
@hanouchen Sure, no problem.

```c
// RUN: %llvmgcc  -g -emit-llvm -O0 -c -o %t.bc %s
// RUN: rm -rf %t.klee-out
// RUN: %klee --output-dir=%t.klee-out --libc=klee --no-output --exit-on-error %t.bc > %t.log

#include "klee/klee.h"
#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

int debug1(int val) {
  int b = 5;
  b += val;
  return b;
}

int main(int argc, char **argv) {
  int easy_to_spot_variable1 = 4;
  int easy_to_spot_variable2 = debug1(argc);
  int i = 0;

  easy_to_spot_variable1 += argc;
  printf("a = %u\n", easy_to_spot_variable1);
  printf("b = %u\n", easy_to_spot_variable2);

  for (i = 0; i < argc*3; ++i) {
    if (i % argc)
      easy_to_spot_variable1 += argc;
  }

  printf("Final a = %u\n", easy_to_spot_variable1);

  //CHECK: test =
  return easy_to_spot_variable2;
}
```

**Comment by hanouchen** [comment]
@MartinNowack Thanks!

**Comment by hanouchen** [comment]
@MartinNowack Hi Martin.  For some reason, the source location provided in the `state->pc->info` is still different from the source location provided in the debug metadata of an`llvm:: Instruction`. For example, I used this code to print the source code of the next instruction of a state:
 ``` 
    auto file = state->pc->info->file;
    auto lineFromInfo = state->pc->info->line;
    auto lineFromDebugLoc = state->pc->inst->getDebugLoc().getLine();
    llvm::outs() << "Line from info: " << lineFromInfo << "\n";
    llvm::outs() << "Line from debugLoc: " << lineFromDebugLoc << "\n";

    std::string source1 = getSourceLine(file, lineFromInfo);
    std::string source2 = getSourceLine(file, lineFromDebugLoc);
    llvm::outs() << "source1: " << source1 << "\n";
    llvm::outs() << "source2: " << source2 << "\n";

    llvm::outs() << "Assembly: ";
    state->pc->inst->dump();
```
And the result I got looked like this:
screen shot 2018-05-13 at 16 48 51

Here, `state->pc->info` is pointing to line 18 and the other one is pointing to line 16. Looking at the assembly, which is an assignment to x (x is defined in a subscope, hence the suffix 1), line 16 seems to be the corresponding source code. Do you why it's behaving like this? Thanks.

**Comment by MartinNowack** [comment]
@hanouchen Please have a look here: https://github.com/hanouchen/klee/commit/4297050c02a27f539c680400d8fd3ea602e2f5d5

I think that's the issue.

HTH.

**Comment by hanouchen** [comment]
@MartinNowack Thanks, that seems to be the problem.

**Pull request merged**